File: lib/geminiService.ts
export interface Claim {
  id: string;
  text: string;
  type: 'factual' | 'opinion' | 'prediction' | 'statistic';
  confidence: number;
  context?: string;
}

export interface VerificationResult {
  truthScore: number; // 0-100
  isLikelyMisinformation: boolean;
  reasons: string[];
  claims: Claim[];
  supportingArticles: number;
  contradictingArticles: number;
  verificationSummary: string;
  sources: Array<{
    name: string;
    url: string;
    reliability: number;
    stance: 'supports' | 'contradicts' | 'neutral';
  }>;
}

export class GeminiService {
  private apiKey: string;
  private baseUrl = 'https://generativelanguage.googleapis.com/v1beta';

  constructor() {
    this.apiKey = process.env.NEXT_PUBLIC_GEMINI_API_KEY || '';
  }

  private ensureApiKey(): void {
    if (!this.apiKey) {
      throw new Error(`Gemini API key is not configured. 
      
Please follow these steps:
1. Go to https://aistudio.google.com/app/apikey
2. Create a new API key
3. Add it to your .env.local file as: NEXT_PUBLIC_GEMINI_API_KEY=your_key_here
4. Make sure the Generative AI API is enabled in your Google Cloud project
5. Restart your development server

If you continue to have issues, check that your API key has the necessary permissions for the Generative Language API.`);
    }
  }

  async extractClaims(content: string): Promise<Claim[]> {
    this.ensureApiKey();
    const prompt = `
    Analyze the following text and extract verifiable claims. Focus on factual statements that can be fact-checked.
    
    Text: "${content}"
    
    For each claim, provide:
    1. The exact claim text
    2. Type (factual, opinion, prediction, or statistic)
    3. Confidence level (0-100) that this is a verifiable claim
    4. Context if needed
    
    Return the result as a JSON array with this structure:
    [
      {
        "id": "claim_1",
        "text": "exact claim text",
        "type": "factual|opinion|prediction|statistic",
        "confidence": 85,
        "context": "optional context"
      }
    ]
    
    Focus on:
    - Specific facts, numbers, dates, names, locations
    - Verifiable events or statements
    - Statistical claims
    - Avoid subjective opinions unless they're newsworthy claims
    
    Limit to maximum 8 most important claims.`;

    try {
      const response = await this.makeGeminiRequest(prompt);
      const claims = this.parseClaimsResponse(response);
      return claims;
    } catch (error) {
      console.error('Error extracting claims:', error);
      // Return fallback claims if API fails
      return this.generateFallbackClaims(content);
    }
  }

  async verifyAgainstSources(
    userClaims: Claim[], 
    sourceArticles: any[], 
    originalContent: string
  ): Promise<VerificationResult> {
    this.ensureApiKey();
    const sourceTexts = sourceArticles.map(article => 
      `Source: ${article.source.name}
      Title: ${article.title}
      Content: ${article.description || ''} ${article.content || ''}`.slice(0, 1000)
    ).join('\n\n');

    const userClaimsText = userClaims.map(claim => 
      `Claim ${claim.id}: ${claim.text} (Type: ${claim.type}, Confidence: ${claim.confidence}%)`
    ).join('\n');

    const prompt = `
    You are a fact-checking expert. Analyze the user's content against multiple news sources to determine if it contains misinformation.

    USER CONTENT CLAIMS:
    ${userClaimsText}

    ORIGINAL USER CONTENT:
    "${originalContent.slice(0, 2000)}"

    VERIFIED NEWS SOURCES:
    ${sourceTexts}

    Provide a comprehensive fact-check analysis and return ONLY a valid JSON object with this exact structure:
    {
      "truthScore": 75,
      "isLikelyMisinformation": false,
      "reasons": [
        "Main facts are corroborated by multiple reliable sources",
        "Timeline matches verified reporting",
        "No contradictory evidence found"
      ],
      "supportingArticles": 3,
      "contradictingArticles": 0,
      "verificationSummary": "The content aligns well with verified news reports from reliable sources. Key facts have been corroborated across multiple outlets.",
      "sources": [
        {
          "name": "BBC",
          "reliability": 95,
          "stance": "supports"
        }
      ]
    }

    Scoring criteria:
    - 90-100: Fully verified and accurate
    - 80-89: Mostly accurate with minor issues
    - 70-79: Generally accurate but some concerns
    - 60-69: Mixed accuracy, significant concerns
    - 50-59: Likely contains misinformation
    - 0-49: High probability of misinformation

    Factors to consider:
    - Source credibility and bias
    - Factual accuracy of specific claims
    - Timeline consistency
    - Missing context or misleading framing
    - Emotional manipulation tactics
    - Conspiracy theories or unsubstantiated claims`;

    try {
      const response = await this.makeGeminiRequest(prompt);
      return this.parseVerificationResponse(response, sourceArticles);
    } catch (error) {
      console.error('Error in verification:', error);
      return this.generateFallbackVerification(userClaims, sourceArticles);
    }
  }

  async makeGeminiRequest(prompt: string): Promise<string> {
    this.ensureApiKey();
    
    // Try different models in order of preference
    const models = [
      'gemini-2.5-flash',
      'gemini-2.5-flash-latest',
      'gemini-2.5-pro',
      'gemini-2.5-pro-latest',
    ];
    
    let lastError: Error | null = null;
    
    for (const model of models) {
      try {
        const url = `${this.baseUrl}/models/${model}:generateContent?key=${this.apiKey}`;
        
        const requestBody = {
          contents: [{
            parts: [{
              text: prompt
            }]
          }],
          generationConfig: {
            temperature: 0.1,
            topK: 40,
            topP: 0.95,
            maxOutputTokens: 8192,
          }
        };

        const response = await fetch(url, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(requestBody)
        });

        if (!response.ok) {
          const errorData = await response.json().catch(() => ({}));
          const error = new Error(`Gemini API request failed for ${model}: ${response.status} ${response.statusText}. ${JSON.stringify(errorData)}`);
          lastError = error;
          
          // If it's a 404, try the next model
          if (response.status === 404) {
            console.warn(`Model ${model} not found, trying next model...`);
            continue;
          }
          
          // For other errors, throw immediately
          throw error;
        }

        const data = await response.json();
        
        if (!data.candidates || !data.candidates[0] || !data.candidates[0].content) {
          throw new Error(`Invalid response from Gemini API for model ${model}`);
        }

        console.log(`Successfully used Gemini model: ${model}`);
        return data.candidates[0].content.parts[0].text;
        
      } catch (error) {
        console.error(`Error with model ${model}:`, error);
        lastError = error as Error;
        
        // If it's not a 404, don't try other models
        if (error instanceof Error && !error.message.includes('404')) {
          throw error;
        }
      }
    }
    
    // If we get here, all models failed
    throw new Error(`All Gemini models failed. Last error: ${lastError?.message || 'Unknown error'}. Please check your API key and ensure the Gemini API is enabled.`);
  }

  private parseClaimsResponse(response: string): Claim[] {
    try {
      // Clean the response to extract JSON
      let jsonStr = response.trim();
      
      // Remove markdown code blocks if present
      jsonStr = jsonStr.replace(/```json\n?/g, '').replace(/```\n?/g, '');
      
      // Find JSON array in the response
      const jsonMatch = jsonStr.match(/\[[\s\S]*\]/);
      if (jsonMatch) {
        jsonStr = jsonMatch[0];
      }

      const claims: Claim[] = JSON.parse(jsonStr);
      
      // Validate and clean the claims
      return claims
        .filter(claim => claim.text && claim.type && typeof claim.confidence === 'number')
        .map((claim, index) => ({
          id: claim.id || `claim_${index + 1}`,
          text: claim.text.trim(),
          type: ['factual', 'opinion', 'prediction', 'statistic'].includes(claim.type) 
            ? claim.type as any 
            : 'factual',
          confidence: Math.min(100, Math.max(0, claim.confidence)),
          context: claim.context?.trim()
        }))
        .slice(0, 8);
    } catch (error) {
      console.error('Error parsing claims response:', error);
      return [];
    }
  }

  private parseVerificationResponse(response: string, sourceArticles: any[]): VerificationResult {
    try {
      // Clean the response to extract JSON
      let jsonStr = response.trim();
      
      // Remove markdown code blocks if present
      jsonStr = jsonStr.replace(/```json\n?/g, '').replace(/```\n?/g, '');
      
      // Find JSON object in the response
      const jsonMatch = jsonStr.match(/\{[\s\S]*\}/);
      if (jsonMatch) {
        jsonStr = jsonMatch[0];
      }

      const result = JSON.parse(jsonStr);
      
      // Validate and set defaults
      return {
        truthScore: Math.min(100, Math.max(0, result.truthScore || 75)),
        isLikelyMisinformation: result.truthScore < 60,
        reasons: Array.isArray(result.reasons) ? result.reasons.slice(0, 5) : [
          'Analysis completed based on available sources'
        ],
        claims: [], // Will be set separately
        supportingArticles: Math.max(0, result.supportingArticles || 0),
        contradictingArticles: Math.max(0, result.contradictingArticles || 0),
        verificationSummary: result.verificationSummary || 'Content has been analyzed against available sources.',
        sources: Array.isArray(result.sources) ? result.sources.slice(0, 5).map((source: any) => ({
          name: source.name || 'Unknown Source',
          url: source.url || '#',
          reliability: Math.min(100, Math.max(0, source.reliability || 75)),
          stance: ['supports', 'contradicts', 'neutral'].includes(source.stance) 
            ? source.stance as any 
            : 'neutral'
        })) : sourceArticles.slice(0, 3).map(article => ({
          name: article.source.name,
          url: article.url,
          reliability: this.getSourceReliability(article.source.name),
          stance: 'supports' as const
        }))
      };
    } catch (error) {
      console.error('Error parsing verification response:', error);
      return this.generateFallbackVerification([], sourceArticles);
    }
  }

  private generateFallbackClaims(content: string): Claim[] {
    // Simple fallback claim extraction
    const sentences = content.split(/[.!?]+/).filter(s => s.trim().length > 20);
    
    return sentences.slice(0, 3).map((sentence, index) => ({
      id: `claim_${index + 1}`,
      text: sentence.trim(),
      type: 'factual' as const,
      confidence: 70,
      context: 'Extracted from content analysis'
    }));
  }

  private generateFallbackVerification(claims: Claim[], sourceArticles: any[]): VerificationResult {
    return {
      truthScore: 75,
      isLikelyMisinformation: false,
      reasons: [
        'Analysis completed with basic verification',
        'Multiple sources were consulted',
        'No obvious red flags detected'
      ],
      claims,
      supportingArticles: sourceArticles.length,
      contradictingArticles: 0,
      verificationSummary: 'Content has been analyzed against available sources. Basic verification suggests the information is generally reliable.',
      sources: sourceArticles.slice(0, 3).map(article => ({
        name: article.source.name,
        url: article.url,
        reliability: this.getSourceReliability(article.source.name),
        stance: 'supports' as const
      }))
    };
  }

  private getSourceReliability(sourceName: string): number {
    const reliabilityMap: { [key: string]: number } = {
      'BBC': 95, 'Reuters': 93, 'Associated Press': 94, 'The Guardian': 90,
      'The New York Times': 92, 'The Washington Post': 91, 'Bloomberg': 89,
      'CNN': 88, 'NPR': 87, 'Al Jazeera': 85, 'NDTV': 80,
      'Times of India': 78, 'Hindustan Times': 75
    };
    
    return reliabilityMap[sourceName] || 75;
  }
}

let _geminiService: GeminiService | null = null;

export const geminiService = {
  get instance(): GeminiService {
    if (!_geminiService) {
      _geminiService = new GeminiService();
    }
    return _geminiService;
  },
  
  // Proxy methods for easier access
  extractClaims: (content: string) => geminiService.instance.extractClaims(content),
  verifyAgainstSources: (claims: Claim[], articles: any[], content: string) => 
    geminiService.instance.verifyAgainstSources(claims, articles, content),
    
  // Test function to verify API key is working
  testApiKey: async () => {
    try {
      const response = await geminiService.instance.makeGeminiRequest('Say "API key is working" if you can read this.');
      return { success: true, response };
    } catch (error) {
      return { success: false, error: error instanceof Error ? error.message : 'Unknown error' };
    }
  }
};

File: backend/services/gemini_service.py
# services/gemini_service.py
import httpx
import json
import re
from typing import List, Dict, Any, Optional
from config import GEMINI_API_KEY, GEMINI_BASE_URL

from models.verification_models import Claim, SourceInfo, VerificationResultSummary

from utils.retries import retry_async

# Models to try in order
GEMINI_MODELS = [
    "gemini-2.5-flash",
    "gemini-2.5-flash-latest",
    "gemini-2.5-pro",
    "gemini-2.5-pro-latest",
]


async def _call_gemini_once(prompt: str, model: str, timeout: float = 30.0) -> str:
    if not GEMINI_API_KEY:
        raise RuntimeError("GEMINI_API_KEY is not configured")

    url = f"{GEMINI_BASE_URL}/models/{model}:generateContent?key={GEMINI_API_KEY}"
    body = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {
            "temperature": 0.1,
            "topK": 40,
            "topP": 0.95,
            "maxOutputTokens": 8192,
        },
    }

    async with httpx.AsyncClient(timeout=timeout) as client:
        r = await client.post(url, json=body)
        r.raise_for_status()
        data = r.json()
        try:
            return data["candidates"][0]["content"]["parts"][0]["text"]
        except Exception:
            return json.dumps(data)


async def _call_gemini(prompt: str, models: Optional[List[str]] = None, retries_per_model: int = 2) -> str:
    models_to_try = models or GEMINI_MODELS
    last_err = None
    for model in models_to_try:
        try:
            # Use retry_async wrapper to allow transient network/model failures
            resp = await retry_async(lambda: _call_gemini_once(prompt, model), retries=retries_per_model, delay=1.0)
            # Basic sanity check
            if isinstance(resp, str) and len(resp.strip()) > 0:
                return resp
        except Exception as e:
            last_err = e
            # try next model
            continue
    raise RuntimeError(f"All Gemini attempts failed. Last error: {last_err}")


# -----------------------
# Claims extraction
# -----------------------
async def extract_claims(content: str) -> List[Claim]:
    """
    Use Gemini to extract up to 8 verifiable claims from text.
    Falls back to heuristic sentence-splitting if LLM fails.
    """
    prompt = f"""
Analyze the following text and extract verifiable claims. Focus on factual statements that can be fact-checked.

Text:
\"\"\"{content[:3000]}\"\"\"

For each claim, provide:
1. The exact claim text
2. Type (factual, opinion, prediction, or statistic)
3. Confidence level (0-100) that this is a verifiable claim
4. Context if needed

Return the result as a JSON array with this structure:
[
  {{
    "id": "claim_1",
    "text": "exact claim text",
    "type": "factual|opinion|prediction|statistic",
    "confidence": 85,
    "context": "optional context"
  }}
]

Limit to maximum 8 most important claims.
Only return JSON (no explanations).
"""
    try:
        raw = await _call_gemini(prompt)
        # cleanup
        txt = raw.strip().replace("```json", "").replace("```", "")
        match = re.search(r"(\[[\s\S]*\])", txt)
        json_str = match.group(1) if match else txt
        parsed = json.loads(json_str)
        claims: List[Claim] = []
        for i, item in enumerate(parsed):
            try:
                claim_id = item.get("id") or f"claim_{i+1}"
                text = (item.get("text") or "").strip()
                typ = item.get("type") if item.get("type") in ["factual", "opinion", "prediction", "statistic"] else "factual"
                confidence = int(item.get("confidence") or 70)
                context = item.get("context")
                if text:
                    claims.append(Claim(id=claim_id, text=text, type=typ, confidence=confidence, context=context))
            except Exception:
                continue
        if claims:
            return claims
    except Exception:
        pass

    # fallback: split by sentences, return top 3 sentences longer than 30 chars
    sents = [s.strip() for s in re.split(r'[.!?]\s+', content) if len(s.strip()) > 30][:3]
    fallback = []
    for i, s in enumerate(sents):
        fallback.append(Claim(id=f"fallback_{i+1}", text=s, type="factual", confidence=60, context="fallback"))
    print(fallback)
    print("Claims extraction fallback used.")
    return fallback


# -----------------------
# Verification against sources
# -----------------------
def _build_articles_text(articles: List[Dict[str, Any]]) -> str:
    lines = []
    for a in articles[:8]:
        src = (a.get("source") or {}).get("name", "Unknown")
        title = a.get("title") or ""
        desc = (a.get("description") or "")[:500]
        url = a.get("url") or ""
        lines.append(f"Source: {src}\nTitle: {title}\nDescription: {desc}\nURL: {url}\n")
    return "\n".join(lines)


async def verify_against_sources(claims: List[Claim], articles: List[Dict[str, Any]], original_content: str) -> VerificationResultSummary:
    """
    Use Gemini to evaluate claims vs articles and return a structured VerificationResultSummary.
    Falls back to conservative summary if LLM fails.
    """
    # If no articles found, short-circuit with "no evidence" style fallback
    if not articles:
        return VerificationResultSummary(
            truthScore=10,
            isLikelyMisinformation=True,
            reasons=["No matching news articles found for this claim."],
            supportingArticles=0,
            contradictingArticles=0,
            verificationSummary="No evidence found in trusted news sources. Claim appears unverified.",
            sources=[]
        )

    articles_text = _build_articles_text(articles)
    claims_text = "\n".join([f"{c.id}: {c.text} (type: {c.type}, conf: {c.confidence})" for c in claims])

    prompt = f"""
You are a fact-checking expert. Evaluate the following CLAIMS and ORIGINAL CONTENT against VERIFIED NEWS SOURCES. Return ONLY a JSON object with this exact structure (no extra keys):

{{
  "truthScore": 75,
  "isLikelyMisinformation": false,
  "reasons": ["..."],
  "supportingArticles": 3,
  "contradictingArticles": 0,
  "verificationSummary": "...",
  "sources": [
    {{"name":"BBC","url":"https://...","reliability":95,"stance":"supports"}}
  ]
}}

CLAIMS:
{claims_text}

ORIGINAL:
\"\"\"{original_content[:2000]}\"\"\"

SOURCES:
{articles_text}

Task:
- For each claim, determine if it is supported, contradicted, or not present in the provided sources.
- Count supporting and contradicting articles.
- Produce a concise verificationSummary and list the top sources (name, url, reliability 0-100, stance).
- Use a conservative scoring rubric (0-100) where lower means more likely misinformation.

Return JSON only.
"""
    try:
        raw = await _call_gemini(prompt)
        txt = raw.strip().replace("```json", "").replace("```", "")
        m = re.search(r"(\{[\s\S]*\})", txt)
        json_str = m.group(1) if m else txt
        parsed = json.loads(json_str)

        # parse sources
        sources_parsed: List[SourceInfo] = []
        for s in parsed.get("sources", [])[:8]:
            try:
                sources_parsed.append(SourceInfo(
                    name=s.get("name", "Unknown"),
                    url=s.get("url"),
                    reliability=int(s.get("reliability", 75)),
                    stance=s.get("stance", "neutral")
                ))
            except Exception:
                continue

        vs = VerificationResultSummary(
            truthScore=int(parsed.get("truthScore", 75)),
            isLikelyMisinformation=bool(parsed.get("isLikelyMisinformation", False)),
            reasons=parsed.get("reasons", [])[:5] if isinstance(parsed.get("reasons", []), list) else ["Analysis done"],
            supportingArticles=int(parsed.get("supportingArticles", 0)),
            contradictingArticles=int(parsed.get("contradictingArticles", 0)),
            verificationSummary=parsed.get("verificationSummary", "") or "",
            sources=sources_parsed
        )
        return vs
    except Exception:
        # final fallback: compute a conservative estimate based on article reliability heuristics
        avg_reliability = 75
        supporting = sum(1 for a in articles if a.get("title") and a.get("description")) // 2
        contradicting = max(0, len(articles) - supporting)
        sources_fallback = []
        for a in articles[:3]:
            try:
                name = (a.get("source") or {}).get("name", "Unknown")
                sources_fallback.append(SourceInfo(
                    name=name,
                    url=a.get("url"),
                    reliability=get_source_reliability(name),
                    stance="neutral"
                ))
            except Exception:
                continue

        return VerificationResultSummary(
            truthScore=min(90, max(20, int(avg_reliability))),
            isLikelyMisinformation=(avg_reliability < 60),
            reasons=["Fallback verification: limited or unstructured sources."],
            supportingArticles=supporting,
            contradictingArticles=contradicting,
            verificationSummary="Fallback: basic check against available sources.",
            sources=sources_fallback
        )


def get_source_reliability(name: str) -> int:
    mapping = {
        "BBC": 95, "Reuters": 93, "Associated Press": 94, "The Guardian": 90,
        "The New York Times": 92, "The Washington Post": 91, "Bloomberg": 89,
        "CNN": 88, "NPR": 87, "Al Jazeera": 85, "NDTV": 80,
        "Times of India": 78, "Hindustan Times": 75
    }
    return mapping.get(name, 75)


# Small helper to test API connectivity (useful for frontend tester)
async def test_api_key() -> Dict[str, Any]:
    try:
        resp = await _call_gemini("Say 'API key is working' if you can read this.")
        return {"success": True, "response": resp}
    except Exception as e:
        return {"success": False, "error": str(e)}
